ddtrace/internal/_queue.pyx

```diff
@@ -0,0 +1,52 @@
+import threading
+
+from ddtrace.vendor import attr
+from ..internal.logger import get_logger
+from . import _rand
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceQueue(object):
+
+    maxsize = attr.ib(type=int, default=0)
+    _lock = attr.ib(init=False, factory=threading.Lock, repr=False)
+    _queue = attr.ib(init=False, factory=list, repr=False)
+    _accepted = attr.ib(init=False, type=int, default=0)
+    _accepted_lengths = attr.ib(init=False, type=int, default=0)
+    _dropped = attr.ib(init=False, type=int, default=0)
+
+    def __len__(self):
+        return len(self._queue)
+
+    def put(self, item):
+        with self._lock:
+            if self.maxsize <= 0 or len(self._queue) < self.maxsize:
+                self._queue.append(item)
+            else:
+                idx = _rand.rand64bits() % len(self._queue)
+                self._queue[idx] = item
+
+                log.warning("Trace queue '%r' is full, dropping a random trace", self)
```

Comment:
What does `repr(TraceQueue())` look like?

`Trace queue 'TraceQueue(maxsize=1000, _accepted=10, _accepted_lengths=100, _dropped=1)' is full, dropping a random trace` ?

Should we increment `_dropped` before we log?

We might be able to do `s/Trace queue '%r'/%r/` here, should be clear (?)

Reply:
Yep that's what it looks like.

:+1: to incrementing and getting rid of the quotes.